From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gellert <gellert.hegyi@around.co>
Date: Mon, 12 Apr 2021 16:45:05 +0200
Subject: adds support to pause and resume audio worklet thread

This patch adds support using the around-dc internal datachannel to pause and resume audio processing on the audio worklet thread.

Signed-off-by: gellert <gellert.hegyi@around.co>

diff --git a/third_party/blink/renderer/platform/audio/audio_destination.cc b/third_party/blink/renderer/platform/audio/audio_destination.cc
index 0cfcf81a327418830f83c5198799c6cdd592540e..e7ec437841854d7a5153f93f1e354fe4ee0b086c 100644
--- a/third_party/blink/renderer/platform/audio/audio_destination.cc
+++ b/third_party/blink/renderer/platform/audio/audio_destination.cc
@@ -213,15 +213,17 @@ void AudioDestination::Render(const WebVector<float*>& destination_data,
     output_bus_->SetChannelMemory(i, destination_data[i], number_of_frames);
 
   if (worklet_task_runner_) {
-    // Use the dual-thread rendering if the AudioWorklet is activated.
-    size_t frames_to_render =
-        fifo_->PullAndUpdateEarmark(output_bus_.get(), number_of_frames);
-    PostCrossThreadTask(
-        *worklet_task_runner_, FROM_HERE,
-        CrossThreadBindOnce(&AudioDestination::RequestRender,
-                            WrapRefCounted(this), number_of_frames,
-                            frames_to_render, delay, delay_timestamp,
-                            prior_frames_skipped));
+    if (!should_pause_worklet_) {
+      // Use the dual-thread rendering if the AudioWorklet is activated.
+      size_t frames_to_render =
+          fifo_->PullAndUpdateEarmark(output_bus_.get(), number_of_frames);
+      PostCrossThreadTask(
+          *worklet_task_runner_, FROM_HERE,
+          CrossThreadBindOnce(&AudioDestination::RequestRender,
+                              WrapRefCounted(this), number_of_frames,
+                              frames_to_render, delay, delay_timestamp,
+                              prior_frames_skipped));
+    }
   } else {
     // Otherwise use the single-thread rendering.
     size_t frames_to_render = fifo_->Pull(output_bus_.get(), number_of_frames);
@@ -324,6 +326,8 @@ void AudioDestination::StartWithWorkletTaskRunner(
   worklet_task_runner_ = std::move(worklet_task_runner);
   web_audio_device_->Start();
   SetDeviceState(DeviceState::kRunning);
+
+  SetupDataChannelObserver();
 }
 
 void AudioDestination::Stop() {
@@ -331,6 +335,8 @@ void AudioDestination::Stop() {
   TRACE_EVENT0("webaudio", "AudioDestination::Stop");
   SendLogMessage(String::Format("%s", __func__));
 
+  TeardownDataChannelObserver();
+
   if (device_state_ == DeviceState::kStopped)
     return;
   web_audio_device_->Stop();
@@ -442,4 +448,10 @@ void AudioDestination::SendLogMessage(const String& message) {
                        .Utf8());
 }
 
+void AudioDestination::OnCommandReceived(std::string command, std::map<std::string, std::string> params) {
+  if (command == "audio_worklet_status") {
+    should_pause_worklet_ = params["paused"] == "true";
+  }
+}
+
 }  // namespace blink
diff --git a/third_party/blink/renderer/platform/audio/audio_destination.h b/third_party/blink/renderer/platform/audio/audio_destination.h
index 2847edda3438ea68e281c5ea926127de36d35206..81627a099a5550aaa828058362fc5fca61a8fd01 100644
--- a/third_party/blink/renderer/platform/audio/audio_destination.h
+++ b/third_party/blink/renderer/platform/audio/audio_destination.h
@@ -43,6 +43,7 @@
 #include "third_party/blink/renderer/platform/wtf/text/wtf_string.h"
 #include "third_party/blink/renderer/platform/wtf/thread_safe_ref_counted.h"
 #include "third_party/blink/renderer/platform/wtf/threading_primitives.h"
+#include "third_party/webrtc/pc/data_channel_controller.h"
 
 namespace blink {
 
@@ -61,7 +62,8 @@ class WebAudioLatencyHint;
 //    AudioWorkletThread.
 class PLATFORM_EXPORT AudioDestination
     : public ThreadSafeRefCounted<AudioDestination>,
-      public WebAudioDevice::RenderCallback {
+      public WebAudioDevice::RenderCallback,
+      public webrtc::AroundCommandObserver {
   USING_FAST_MALLOC(AudioDestination);
 
  public:
@@ -137,6 +139,8 @@ class PLATFORM_EXPORT AudioDestination
   // This should only be called from the audio thread.
   unsigned RenderQuantumFrames() const { return render_quantum_frames_; }
 
+  void OnCommandReceived(std::string command, std::map<std::string, std::string> params) override;
+
  private:
   void SetDeviceState(DeviceState);
 
@@ -196,6 +200,8 @@ class PLATFORM_EXPORT AudioDestination
   // Modified only on the main thread, so it can be read without holding a lock
   // there.
   DeviceState device_state_;
+
+  bool should_pause_worklet_ = false;
 };
 
 }  // namespace blink
